simulated_beacon/simulated_beacon.go

```diff
@@ -0,0 +1,236 @@
+			HeadBlockHash:      header.Hash(),
+			SafeBlockHash:      header.Hash(),
+			FinalizedBlockHash: header.Hash(),
+		}
+	)
+
+	// if genesis block, send forkchoiceUpdated to trigger transition to PoS
+	if header.Number.BitLen() == 0 {
+		if _, err := engineAPI.ForkchoiceUpdatedV2(curForkchoiceState, nil); err != nil {
+			log.Crit("failed to initiate PoS transition for genesis via Forkchoiceupdated", "err", err)
+		}
+	}
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			break
+		case curTime := <-ticker.C:
+			if curTime.Unix() > lastBlockTime.Add(c.period).Unix() {
+				c.mu.Lock()
+				feeRecipient := c.feeRecipient
+				pendingWithdrawals := c.withdrawals.pop()
+				c.mu.Unlock()
+
+				payloadAttr := &engine.PayloadAttributes{
+					Timestamp:             uint64(curTime.Unix()),
+					Random:                common.Hash{}, // TODO: make this configurable?
+					SuggestedFeeRecipient: feeRecipient,
+					Withdrawals:           pendingWithdrawals,
+				}
+
+				// trigger block building
+				fcState, err := engineAPI.ForkchoiceUpdatedV2(curForkchoiceState, payloadAttr)
+				if err != nil {
+					log.Crit("failed to trigger block building via forkchoiceupdated", "err", err)
+				}
+
+				var payload *engine.ExecutableData
+
+				var (
+					restartPayloadBuilding bool
+					// building a payload times out after SECONDS_PER_SLOT (12s on mainnet).
+					// trigger building a new payload if this amount of time elapses w/o any transactions or withdrawals
+					// having been received.
+					payloadTimeout = time.NewTimer(12 * time.Second)
+					// interval to poll the pending state to detect if transactions have arrived, and proceed if they have
+					// (or if there are pending withdrawals to include)
+					buildTicker = time.NewTicker(100 * time.Millisecond)
+				)
+				for {
+					select {
+					case <-buildTicker.C:
+						pendingHeader, err := c.eth.APIBackend.HeaderByNumber(context.Background(), rpc.PendingBlockNumber)
+						if err != nil {
+							log.Crit("failed to get pending block header", "err", err)
+						}
+						// don't build a block if we don't have pending txs or withdrawals
+						if pendingHeader.TxHash == types.EmptyTxsHash && len(payloadAttr.Withdrawals) == 0 {
+							continue
+						}
+						payload, err = engineAPI.GetPayloadV1(*fcState.PayloadID)
+						if err != nil {
+							log.Crit("error retrieving payload", "err", err)
+						}
+						// Don't build a block if it doesn't contain transactions or withdrawals.
+						// Somehow, txs can arrive, be detected by this routine, but be missed by the miner payload builder.
+						// So this last clause prevents empty blocks from being built.
+						if len(payload.Transactions) == 0 && len(payloadAttr.Withdrawals) == 0 {
+							restartPayloadBuilding = true
+						}
+					case <-payloadTimeout.C:
+						restartPayloadBuilding = true
+					case <-c.ctx.Done():
+						return
+					}
+					break
+				}
+				if restartPayloadBuilding {
+					continue
+				}
```

Comment:
I think this entire block can be simplified into 
```suggestion
			time.Sleep(200 * time.Millisecond) // Give it some time to build
			if payload, err = engineAPI.GetPayloadV1(*fcState.PayloadID); err != nil {
				log.Crit("error retrieving payload", "err", err)
			}
			// Don't build a block if it doesn't contain transactions or withdrawals.
			if payload == nil || len(payload.Transactions) == 0 && len(payloadAttr.Withdrawals) == 0 {
				continue
			}
```

Reply:
This approach of having a static payload building time is something I discussed with @fjl .  It is simpler than the current approach.  However, the downside is that if users want to configure their dev chain with some incredibly large gas limit and execute very expensive transactions, they will not be included if they take too long to execute.

Reply:
Looks to me like your code is basically the same, but uses `100ms` instead of `200ms` ... ?